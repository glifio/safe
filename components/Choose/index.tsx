import { useCallback, useEffect, useMemo, useState } from 'react'
import { useRouter } from 'next/router'
import {
  getQueryParam,
  navigate,
  ButtonRowSpaced,
  ButtonV2,
  InputV2,
  Dialog,
  ErrorBox,
  ShadowBox,
  SmartLink,
  LoadingScreen
} from '@glif/react-components'

import { useMsig } from '../../MsigProvider'
import { PAGE, QPARAM } from '../../constants'

export const Choose = () => {
  const router = useRouter()
  const { setMsigActor, loading, errors, ActorCode } = useMsig()
  const [msigAddressParam, setMsigAddressParam] = useState<string>(
    getQueryParam.string(router, QPARAM.MSIG_ADDRESS)
  )

  // Input states
  const [safeID, setSafeID] = useState<string>(msigAddressParam)
  const [isSafeIDValid, setIsSafeIDValid] = useState<boolean>(false)
  const [submittedForm, setSubmittedForm] = useState<boolean>(false)
  const [navigating, setNavigating] = useState<boolean>(false)

  // Get error message from MSIG provider
  const errorMessage = useMemo<string>(() => {
    if (errors.actorNotFound) return 'Safe not found'
    if (errors.connectedWalletNotMsigSigner)
      return 'Your wallet is not an owner of this Safe. Please go back and choose a wallet address that is an owner of this Safe.'
    if (errors.notMsigActor)
      return 'The address you entered is not a Glif Safe.'
    return errors.unhandledError || ''
  }, [
    errors.actorNotFound,
    errors.connectedWalletNotMsigSigner,
    errors.notMsigActor,
    errors.unhandledError
  ])

  const submitForm = useCallback(() => {
    setSubmittedForm(true)
    setMsigActor(safeID)
  }, [setSubmittedForm, setMsigActor, safeID])

  // Automatically submit the form if there is
  // a valid MSIG address in the query params
  useEffect(
    () => msigAddressParam && !submittedForm && submitForm(),
    [msigAddressParam, submittedForm, submitForm]
  )

  // Clear the MSIG address retrieved from the query params
  // if it failed to load, which removes the loading screen
  useEffect(
    () =>
      submittedForm &&
      errorMessage &&
      msigAddressParam &&
      setMsigAddressParam(''),
    [submittedForm, errorMessage, msigAddressParam, setMsigAddressParam]
  )

  // When there is an ActorCode we successfully retrieved
  // the multisig and we push the user to the msig home
  useEffect(() => {
    if (submittedForm && !errorMessage && !!ActorCode) {
      setNavigating(true)
      navigate(router, { pageUrl: PAGE.MSIG_HOME })
    }
  }, [submittedForm, errorMessage, ActorCode, router])

  return msigAddressParam || loading || navigating ? (
    <LoadingScreen />
  ) : (
    <Dialog>
      <form
        onSubmit={(e) => {
          e.preventDefault()
          submitForm()
        }}
      >
        {submittedForm && errorMessage && <ErrorBox>{errorMessage}</ErrorBox>}
        <ShadowBox>
          <h2>Safe ID</h2>
          <hr />
          <InputV2.Address
            name='safe-id'
            label='Please enter your Safe ID below to continue'
            vertical
            centered
            autoFocus
            truncate={false}
            actor
            value={safeID}
            onChange={setSafeID}
            setIsValid={setIsSafeIDValid}
            disabled={loading}
          />
          <p>
            Don&apos;t have a Safe ID?{' '}
            <SmartLink href={PAGE.MSIG_CREATE}>Create one</SmartLink>
          </p>
        </ShadowBox>
        <ButtonRowSpaced>
          <ButtonV2 large type='button' onClick={() => router.back()}>
            Back
          </ButtonV2>
          <ButtonV2
            large
            green
            type='submit'
            disabled={!isSafeIDValid || loading}
          >
            Submit
          </ButtonV2>
        </ButtonRowSpaced>
      </form>
    </Dialog>
  )
}
