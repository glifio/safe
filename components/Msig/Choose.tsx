import { useEffect, useMemo, useState } from 'react'
import { useRouter } from 'next/router'
import {
  navigate,
  generateRouteWithRequiredUrlParams,
  ButtonRowSpaced,
  ButtonV2,
  InputV2,
  Dialog,
  ErrorBox,
  ShadowBox,
  SmartLink
} from '@glif/react-components'

import { useMsig } from '../../MsigProvider'
import { PAGE } from '../../constants'

export const Choose = () => {
  const router = useRouter()
  const { setMsigActor, errors, ActorCode } = useMsig()

  // Input states
  const [safeID, setSafeID] = useState<string>('')
  const [isSafeIDValid, setIsSafeIDValid] = useState<boolean>(false)
  const [submittedForm, setSubmittedForm] = useState<boolean>(false)

  // Get error message from MSIG provider
  const errorMessage = useMemo(() => {
    if (errors.actorNotFound) return 'Safe not found'
    if (errors.connectedWalletNotMsigSigner)
      return 'Your wallet is not an owner of this Safe. Please go back and choose a wallet address that is an owner of this Safe.'
    if (errors.notMsigActor)
      return 'The address you entered is not a Glif Safe.'
    return errors.unhandledError || ''
  }, [
    errors.actorNotFound,
    errors.connectedWalletNotMsigSigner,
    errors.notMsigActor,
    errors.unhandledError
  ])

  // When there is an ActorCode we successfully retrieved
  // the multisig and we push the user to the msig home
  useEffect(
    () =>
      submittedForm &&
      !errorMessage &&
      !!ActorCode &&
      navigate(router, { pageUrl: PAGE.MSIG_HOME }),
    [submittedForm, errorMessage, ActorCode, router]
  )

  return (
    <Dialog>
      {submittedForm && errorMessage && <ErrorBox>{errorMessage}</ErrorBox>}
      <ShadowBox>
        <h2>Safe ID</h2>
        <hr />
        <form>
          <InputV2.Address
            name='safe-id'
            label='Please enter your Safe ID below to continue'
            vertical
            centered
            autoFocus
            truncate={false}
            actor
            value={safeID}
            onChange={setSafeID}
            setIsValid={setIsSafeIDValid}
          />
        </form>
        <p>
          Don&apos;t have a Safe ID?{' '}
          <SmartLink
            href={generateRouteWithRequiredUrlParams({
              pageUrl: PAGE.MSIG_CREATE,
              existingQParams: {}
            })}
          >
            Create one
          </SmartLink>
        </p>
      </ShadowBox>
      <ButtonRowSpaced>
        <ButtonV2 large onClick={() => router.back()}>
          Back
        </ButtonV2>
        <ButtonV2
          large
          green
          disabled={!isSafeIDValid}
          onClick={() => {
            setSubmittedForm(true)
            setMsigActor(safeID)
          }}
        >
          Submit
        </ButtonV2>
      </ButtonRowSpaced>
    </Dialog>
  )
}
