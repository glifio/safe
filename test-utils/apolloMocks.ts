import { AddressDocument } from '@glif/react-components'
import {
  MULTISIG_SIGNER_ADDRESS,
  MULTISIG_SIGNER_ADDRESS_2,
  MULTISIG_SIGNER_ID,
  MULTISIG_SIGNER_ID_2
} from './constants'

export const addressMocks = [
  {
    request: {
      query: AddressDocument,
      variables: {
        address: MULTISIG_SIGNER_ID
      }
    },
    result: {
      data: {
        address: {
          robust: '',
          id: MULTISIG_SIGNER_ID
        }
      }
    }
  },
  {
    request: {
      query: AddressDocument,
      variables: {
        address: MULTISIG_SIGNER_ID_2
      }
    },
    result: {
      data: {
        address: {
          robust: '',
          id: MULTISIG_SIGNER_ID_2
        }
      }
    }
  },
  {
    request: {
      query: AddressDocument,
      variables: {
        address: MULTISIG_SIGNER_ADDRESS
      }
    },
    result: {
      data: {
        address: {
          robust: MULTISIG_SIGNER_ADDRESS,
          id: MULTISIG_SIGNER_ID
        }
      }
    }
  },
  {
    request: {
      query: AddressDocument,
      variables: {
        address: MULTISIG_SIGNER_ADDRESS_2
      }
    },
    result: {
      data: {
        address: {
          robust: MULTISIG_SIGNER_ADDRESS_2,
          id: MULTISIG_SIGNER_ID_2
        }
      }
    }
  }
]
