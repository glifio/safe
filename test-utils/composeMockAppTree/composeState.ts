import { FilecoinNumber } from '@glif/filecoin-number'
import { createPath, WalletProviderState } from '@glif/react-components'

import { IMPORT_MNEMONIC, IMPORT_SINGLE_KEY } from '../../constants'
import { mockWalletProviderInstance } from '../../__mocks__/@glif/filecoin-wallet-provider'
import { emptyMsigState } from '../../MsigProvider/types'
import {
  WALLET_ADDRESS,
  MULTISIG_ACTOR_ADDRESS,
  signers,
  WALLET_ID,
  WALLET_ADDRESS_2,
  WALLET_ID_2
} from '../constants'

export type Preset =
  | 'preOnboard'
  | 'postOnboard'
  | 'postOnboardLowBal'
  | 'postOnboardWithError'
  | 'selectedOtherWallet'
  | 'pendingMsigCreate'

export const composeWalletProviderState = (
  initialWalletProviderState: WalletProviderState,
  preset: Preset
) => {
  switch (preset) {
    case 'pendingMsigCreate':
    case 'postOnboard': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletType: IMPORT_MNEMONIC,
        walletProvider: mockWalletProviderInstance,
        wallets: [
          {
            robust: WALLET_ADDRESS,
            id: WALLET_ID,
            balance: new FilecoinNumber('1', 'fil'),
            path: createPath(1, 0)
          }
        ],
        selectedWalletIdx: 0,
        loginOption: IMPORT_SINGLE_KEY
      })
    }
    case 'postOnboardLowBal': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletType: IMPORT_MNEMONIC,
        walletProvider: mockWalletProviderInstance,
        wallets: [
          {
            robust: WALLET_ADDRESS,
            id: WALLET_ID,
            balance: new FilecoinNumber('.000001', 'fil'),
            path: createPath(1, 0)
          }
        ],
        selectedWalletIdx: 0,
        loginOption: IMPORT_SINGLE_KEY
      })
    }
    case 'postOnboardWithError': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletProvider: mockWalletProviderInstance,
        wallets: [
          {
            robust: WALLET_ADDRESS,
            id: WALLET_ID,
            balance: new FilecoinNumber('1', 'fil'),
            path: createPath(1, 0)
          }
        ],
        selectedWalletIdx: 0,
        loginOption: IMPORT_SINGLE_KEY
      })
    }
    case 'selectedOtherWallet': {
      return Object.freeze({
        ...initialWalletProviderState,
        walletProvider: mockWalletProviderInstance,
        wallets: [
          {
            robust: WALLET_ADDRESS,
            id: WALLET_ID,
            balance: new FilecoinNumber('1', 'fil'),
            path: createPath(1, 0)
          },
          {
            robust: WALLET_ADDRESS_2,
            id: WALLET_ID_2,
            balance: new FilecoinNumber('5', 'fil'),
            path: createPath(1, 1)
          }
        ],
        selectedWalletIdx: 1,
        loginOption: IMPORT_MNEMONIC
      })
    }
    default:
      return initialWalletProviderState
  }
}

export const mockMsigProviderContext = {
  Address: MULTISIG_ACTOR_ADDRESS,
  ActorCode: 'fil/5/multisig',
  Balance: new FilecoinNumber('1', 'fil'),
  AvailableBalance: new FilecoinNumber('1', 'fil'),
  Signers: signers,
  InitialBalance: new FilecoinNumber('1', 'fil'),
  NextTxnID: 0,
  NumApprovalsThreshold: 2,
  StartEpoch: 0,
  UnlockDuration: 0,
  errors: {
    notMsigActor: false,
    connectedWalletNotMsigSigner: false,
    actorNotFound: false,
    unhandledError: ''
  },
  loading: false,
  setMsigActor: null
}

export const composeMsigProviderState = (preset: Preset) => {
  switch (preset) {
    case 'preOnboard': {
      return Object.freeze(emptyMsigState)
    }
    case 'pendingMsigCreate': {
      return Object.freeze(emptyMsigState)
    }
    default: {
      return Object.freeze(mockMsigProviderContext)
    }
  }
}
