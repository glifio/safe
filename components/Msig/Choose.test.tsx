import {
  render,
  act,
  fireEvent,
  getByText,
  getByRole,
  RenderResult
} from '@testing-library/react'

import composeMockAppTree from '../../test-utils/composeMockAppTree'
import { flushPromises } from '../../test-utils'
import { Choose } from './Choose'

const validSafeID = 'f2yrhsjwuwypy4gsv2lcikphizkusqe2gp3pp4w5q'
const inValidSafeID = 't1iuryu3ke2hewrcxp4ezhmr5cmfeq3wjhpxaucza'

describe('Choose', () => {
  test('it renders the initial state correctly', async () => {
    const { Tree } = composeMockAppTree('postOnboard')
    let result: RenderResult | null = null

    await act(async () => {
      result = render(
        <Tree>
          <Choose />
        </Tree>
      )
    })

    // Check whether the submit button is disabled
    const submit = getByText(result.container, 'Submit')
    expect(submit).toBeDisabled()

    // Check snapshot
    expect(result.container.firstChild).toMatchSnapshot()
  })

  test('it renders correctly after entering a valid safe id', async () => {
    const { Tree } = composeMockAppTree('postOnboard')
    let result: RenderResult | null = null

    await act(async () => {
      result = render(
        <Tree>
          <Choose />
        </Tree>
      )

      // Enter a valid Safe ID
      const safeId = getByRole(result.container, 'textbox')
      fireEvent.change(safeId, { target: { value: validSafeID } })
      safeId.blur()

      await flushPromises()
    })

    // Check whether the submit button is enabled
    const submit = getByText(result.container, 'Submit')
    expect(submit).toBeEnabled()

    // Check snapshot
    expect(result.container.firstChild).toMatchSnapshot()
  })

  test('it renders correctly after entering an invalid safe id', async () => {
    const { Tree } = composeMockAppTree('postOnboard')
    let result: RenderResult | null = null

    await act(async () => {
      result = render(
        <Tree>
          <Choose />
        </Tree>
      )

      // Enter a valid Safe ID
      const safeId = getByRole(result.container, 'textbox')
      fireEvent.change(safeId, { target: { value: inValidSafeID } })
      safeId.blur()

      await flushPromises()
    })

    // Check whether the safe id input has an error
    const safeId = getByRole(result.container, 'textbox')
    expect(safeId).toHaveClass('error')

    // Check whether the submit button is disabled
    const submit = getByText(result.container, 'Submit')
    expect(submit).toBeDisabled()

    // Check snapshot
    expect(result.container.firstChild).toMatchSnapshot()
  })
})
