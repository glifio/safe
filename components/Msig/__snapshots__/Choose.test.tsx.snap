// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Choose it renders correctly after entering a valid safe id 1`] = `
.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  font-size: 1.25em;
}

.c6 svg path {
  stroke: var(--black);
}

.c6:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c6:hover svg path {
  stroke: var(--purple-medium);
}

.c6:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c6:active svg path {
  stroke: var(--white) !important;
}

.c7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  font-size: 1.25em;
  color: var(--white);
  border-color: var(--green-medium);
  background: var(--green-medium);
}

.c7 svg path {
  stroke: var(--black);
}

.c7:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c7:hover svg path {
  stroke: var(--purple-medium);
}

.c7:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c7:active svg path {
  stroke: var(--white) !important;
}

.c7 svg path {
  stroke: var(--white);
}

.c7:hover {
  color: var(--green-medium);
  border-color: var(--green-medium);
  background: var(--white);
}

.c7:hover svg path {
  stroke: var(--green-medium);
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  text-align: left;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 * {
  text-align: center;
}

.c3 span.info,
.c3 span.error {
  font-size: 0.875rem;
}

.c3 span.error {
  color: var(--red-medium);
}

.c3 .text-input-wrapper {
  position: relative;
}

.c3 .text-input-wrapper input {
  width: 100%;
}

.c3 .text-input-wrapper .unit {
  position: absolute;
  top: 50%;
  right: 1em;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  color: var(--purple-medium);
}

.c1 {
  padding: 1.5em;
  border-radius: 8px;
  overflow: hidden;
  text-align: center;
  word-break: break-word;
}

.c1 > *:first-child {
  margin-top: 0;
}

.c1 > *:last-child {
  margin-bottom: 0;
}

.c1 > svg:first-child {
  display: block;
  margin: 0 auto 1em;
}

.c1 > h2:first-child {
  margin: 0;
}

.c1 > header:first-child {
  margin: -1.5em -1.5em 1.5em -1.5em;
  padding: 1.5em;
  background-color: var(--purple-medium);
  color: var(--white);
}

.c2 {
  background-color: var(--white);
  box-shadow: 0 0 0.5em var(--gray-light);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  margin: 1.5em 0;
}

.c5 {
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
  width: 100%;
  max-width: 35em;
}

.c0 form {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
}

<div
  class="c0"
>
  <div
    class="c1 c2"
  >
    <h2>
      Safe ID
    </h2>
    <hr />
    <form>
      <label
        class="c3"
      >
        <span>
          Please enter your Safe ID below to continue
        </span>
        
        <div
          class="text-input-wrapper"
        >
          <input
            class=""
            max=""
            min=""
            name="safe-id"
            placeholder="f2..."
            step=""
            style="padding-right: 1em;"
            type="text"
            value="f2yrhsjwuwypy4gsv2lcikphizkusqe2gp3pp4w5q"
          />
          
        </div>
        
      </label>
    </form>
    <p>
      Don't have a Safe ID?
       
      <a
        href="/create"
      >
        Create one
      </a>
    </p>
  </div>
  <div
    class="c4 c5"
  >
    <button
      class="c6"
    >
      Back
    </button>
    <button
      class="c7"
    >
      Submit
    </button>
  </div>
</div>
`;

exports[`Choose it renders correctly after entering an invalid safe id 1`] = `
.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  font-size: 1.25em;
}

.c6 svg path {
  stroke: var(--black);
}

.c6:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c6:hover svg path {
  stroke: var(--purple-medium);
}

.c6:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c6:active svg path {
  stroke: var(--white) !important;
}

.c7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  font-size: 1.25em;
  color: var(--white);
  border-color: var(--green-medium);
  background: var(--green-medium);
  pointer-events: none;
  color: var(--gray-medium);
  border-color: var(--gray-medium);
  background: transparent;
}

.c7 svg path {
  stroke: var(--black);
}

.c7:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c7:hover svg path {
  stroke: var(--purple-medium);
}

.c7:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c7:active svg path {
  stroke: var(--white) !important;
}

.c7 svg path {
  stroke: var(--white);
}

.c7:hover {
  color: var(--green-medium);
  border-color: var(--green-medium);
  background: var(--white);
}

.c7:hover svg path {
  stroke: var(--green-medium);
}

.c7 svg path {
  stroke: var(--gray-medium);
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  text-align: left;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 * {
  text-align: center;
}

.c3 span.info,
.c3 span.error {
  font-size: 0.875rem;
}

.c3 span.error {
  color: var(--red-medium);
}

.c3 .text-input-wrapper {
  position: relative;
}

.c3 .text-input-wrapper input {
  width: 100%;
}

.c3 .text-input-wrapper .unit {
  position: absolute;
  top: 50%;
  right: 1em;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  color: var(--purple-medium);
  color: var(--red-dark) !important;
}

.c1 {
  padding: 1.5em;
  border-radius: 8px;
  overflow: hidden;
  text-align: center;
  word-break: break-word;
}

.c1 > *:first-child {
  margin-top: 0;
}

.c1 > *:last-child {
  margin-bottom: 0;
}

.c1 > svg:first-child {
  display: block;
  margin: 0 auto 1em;
}

.c1 > h2:first-child {
  margin: 0;
}

.c1 > header:first-child {
  margin: -1.5em -1.5em 1.5em -1.5em;
  padding: 1.5em;
  background-color: var(--purple-medium);
  color: var(--white);
}

.c2 {
  background-color: var(--white);
  box-shadow: 0 0 0.5em var(--gray-light);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  margin: 1.5em 0;
}

.c5 {
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
  width: 100%;
  max-width: 35em;
}

.c0 form {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
}

<div
  class="c0"
>
  <div
    class="c1 c2"
  >
    <h2>
      Safe ID
    </h2>
    <hr />
    <form>
      <label
        class="c3"
      >
        <span>
          Please enter your Safe ID below to continue
        </span>
        
        <div
          class="text-input-wrapper"
        >
          <input
            class="error"
            max=""
            min=""
            name="safe-id"
            placeholder="f2..."
            step=""
            style="padding-right: 1em;"
            type="text"
            value="t1iuryu3ke2hewrcxp4ezhmr5cmfeq3wjhpxaucza"
          />
          
        </div>
        <span
          class="error"
        >
          Second character must be 0 or 2
        </span>
      </label>
    </form>
    <p>
      Don't have a Safe ID?
       
      <a
        href="/create"
      >
        Create one
      </a>
    </p>
  </div>
  <div
    class="c4 c5"
  >
    <button
      class="c6"
    >
      Back
    </button>
    <button
      class="c7"
      disabled=""
    >
      Submit
    </button>
  </div>
</div>
`;

exports[`Choose it renders the initial state correctly 1`] = `
.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  font-size: 1.25em;
}

.c6 svg path {
  stroke: var(--black);
}

.c6:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c6:hover svg path {
  stroke: var(--purple-medium);
}

.c6:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c6:active svg path {
  stroke: var(--white) !important;
}

.c7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  font-size: 1.25em;
  color: var(--white);
  border-color: var(--green-medium);
  background: var(--green-medium);
  pointer-events: none;
  color: var(--gray-medium);
  border-color: var(--gray-medium);
  background: transparent;
}

.c7 svg path {
  stroke: var(--black);
}

.c7:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c7:hover svg path {
  stroke: var(--purple-medium);
}

.c7:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c7:active svg path {
  stroke: var(--white) !important;
}

.c7 svg path {
  stroke: var(--white);
}

.c7:hover {
  color: var(--green-medium);
  border-color: var(--green-medium);
  background: var(--white);
}

.c7:hover svg path {
  stroke: var(--green-medium);
}

.c7 svg path {
  stroke: var(--gray-medium);
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  text-align: left;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 * {
  text-align: center;
}

.c3 span.info,
.c3 span.error {
  font-size: 0.875rem;
}

.c3 span.error {
  color: var(--red-medium);
}

.c3 .text-input-wrapper {
  position: relative;
}

.c3 .text-input-wrapper input {
  width: 100%;
}

.c3 .text-input-wrapper .unit {
  position: absolute;
  top: 50%;
  right: 1em;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  color: var(--purple-medium);
}

.c1 {
  padding: 1.5em;
  border-radius: 8px;
  overflow: hidden;
  text-align: center;
  word-break: break-word;
}

.c1 > *:first-child {
  margin-top: 0;
}

.c1 > *:last-child {
  margin-bottom: 0;
}

.c1 > svg:first-child {
  display: block;
  margin: 0 auto 1em;
}

.c1 > h2:first-child {
  margin: 0;
}

.c1 > header:first-child {
  margin: -1.5em -1.5em 1.5em -1.5em;
  padding: 1.5em;
  background-color: var(--purple-medium);
  color: var(--white);
}

.c2 {
  background-color: var(--white);
  box-shadow: 0 0 0.5em var(--gray-light);
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  margin: 1.5em 0;
}

.c5 {
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
  width: 100%;
  max-width: 35em;
}

.c0 form {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 10px;
}

<div
  class="c0"
>
  <div
    class="c1 c2"
  >
    <h2>
      Safe ID
    </h2>
    <hr />
    <form>
      <label
        class="c3"
      >
        <span>
          Please enter your Safe ID below to continue
        </span>
        
        <div
          class="text-input-wrapper"
        >
          <input
            class=""
            max=""
            min=""
            name="safe-id"
            placeholder="f2..."
            step=""
            style="padding-right: 1em;"
            type="text"
            value=""
          />
          
        </div>
        
      </label>
    </form>
    <p>
      Don't have a Safe ID?
       
      <a
        href="/create"
      >
        Create one
      </a>
    </p>
  </div>
  <div
    class="c4 c5"
  >
    <button
      class="c6"
    >
      Back
    </button>
    <button
      class="c7"
      disabled=""
    >
      Submit
    </button>
  </div>
</div>
`;
